File: test/dirwalker.js
import fs from 'fs-extra';
import path from 'path';

generate_dir('/home/ht/test', 'desttt', {});

function generate_dir(src, dst, config) {
  let src_dir = path.normalize(src),
      dst_dir = path.normalize(dst);

  fs.ensureDirSync(dst_dir);
  fs.readdirSync(src_dir).forEach(generate);

  function generate(file) {
    if (!skip(file)) {
      let full_src_path = path.join(src_dir, file),
          full_dst_path = path.join(dst_dir, file);

      if (fs.statSync(full_src_path).isDirectory()) {
        generate_dir(full_src_path, full_dst_path, config);
      } else {
        if (convert(file)) {
          convert_file(full_src_path, full_dst_path, config);
        } else {
          fs.copySync(full_src_path, full_dst_path)
        }
      }
    }
  }

  function skip(file) {
    // All unix-style hidden files and directories are skipped
    return /^\..*$/.test(file);
  }

  function convert(file) {
    // All and only markdown files are converted
    return /$\.(markdown|md)$/.test(path.extname(file));
  }

}

function convert_file(src, dst, config) {
  consolel.log(`Convert ${src} into ${dst} using ${config}`);
}

File: test/dirwalker.rb
require 'pathname'
require 'fileutils'

def generate_dir(src, dst, config) 

  destination = ensure_destination dst
  config = reconfigure config, src
  source_tree = Pathname.new src
  source_tree.each_child do |source_child|
    next if skip? source_child
    
    destination_child = destination.join source_child.basename

    if source_child.directory?
      generate_dir source_child, destination_child, config
    else
      if convert? source_child
        convert_file source_child, destination_child, config
      else
        FileUtils.cp source_child, destination_child
      end
    end
  end

end

def skip?(child)
  child.basename.to_s =~ /^\..*$/
end

def convert?(child)
  child.extname.to_s =~ /^(markdown|md)$/
end

def ensure_destination(dst)
  destination = Pathname.new(dst)
  destination.mkdir if not destination.exist?
  raise "Destination problem" if not destination.directory?
  destination
end


def reconfigure(config, src)
  config
end

generate_dir('/home/ht/test/src', '/home/ht/test/www', true)
